### ISSUE TITLE
investigate validating HAWK header before reading request body

### ISSUE THREAD

**Issue by g-k** [mozilla-services/autograph#114]
From working on #111 https://github.com/mozilla-services/autograph/pull/113, we should try reading and validating the parts of the HTTP header before the whole response.

**Comment by g-k** [comment]
We could also set a global timeout in the http.Server config.

**Comment by g-k** [comment]
So this change would be: 

- [ ] split `authorize` into authorizeHeader and authorizeBody
- [ ] get `r.Header.Get("Authorization")` before the body ReadAll, call authorizeHeader on it, and return early on error
- [ ] if header authorization passes, read the body and call authorize/verify body/payload

For on the auth header we can check:

- [ ] it is not empty: https://github.com/mozilla-services/autograph/blob/15d03e9347c684840597b14146d6fa7b22ad7b93/authorize.go#L31
- [ ] it parses / has the right format: https://github.com/mozilla-services/autograph/blob/15d03e9347c684840597b14146d6fa7b22ad7b93/authorize.go#L34
- [ ] check NewAuthFromRequest errs (creds exist, nonce replay)
- [ ] check ts skew and mac with `auth.Valid()`

And for the body we just need to check:

- [ ] `auth.ValidHash` on the payload hash https://github.com/mozilla-services/autograph/blob/15d03e9347c684840597b14146d6fa7b22ad7b93/authorize.go#L47-L49
- [ ] look into whether the `payloadhash.Write(body)` is streamed or buffered in memory

**Comment by g-k** [comment]
I believe Benson and :alm ran into in an issue where nginx expected the rest of the body to be read from the connection to finish the request, so we still might need to read and discard the rest of the body after an invalid auth header.

https://github.com/mozilla-services/go-syncstorage/blob/e94dd1ddf4c7e0499ebe085f42415c2290902fc2/web/misc.go#L399-L401

@milescrabill are we running behind nginx?

**Comment by milescrabill** [comment]
Yes, we are behind nginx. See: https://github.com/mozilla-services/cloudops-deployment/blob/master/projects/autograph/puppet/modules/autograph/manifests/app.pp#L38-L45

**Comment by g-k** [comment]
OK we'll need to pipe the rest of req body to `/dev/null` then:

```go
if req.Body != nil {
		io.Copy(ioutil.Discard, req.Body)
		req.Body.Close()
}
```

Pretty much all current tests should pass, but we additionally expect:

- [ ] the difference in time between a request with an invalid hawk TS and no body and one with a large body should be smaller (compared to current master)

@ajvb are you free to pick this up?

**Comment by ajvb** [comment]
@g-k yeah, absolutely.

**Comment by g-k** [comment]
Sweet! lmk if you have questions.

**Comment by g-k** [comment]
NB: This might conflict with adding metrics in #116 OTOH getting metrics first will let us know if this improves things

**Comment by g-k** [comment]
Landed the metrics changes. We'll probably want to metric counters for # of requests and # of requests that can return early with this change.
